### PR TITLE
Startup speedups (XML doc cache, delay non-essential tool param parsing & indexing)

### PR DESCRIPTION
**Pull request by mvdbeek**
The new watermark for **full app startup** with main's full toolbox from CVMFS in 9 - 11 seconds (down from 55 - 57 on the dev branch). 

There are three ideas in here: 
- First we store a string with the expanded version of the tool xml document in a dogpile cache (and we override the default pickling since pickling is dead-slow).
- Then we parse tool inputs and outputs only when needed (i.e building to tool form / module for the first time)
- Push toolbox indexing to the background

things left to do:
- ~~move the cache decorator to own module~~
- ~~the modification time needs to be included in the cache key~~
- ~~separate the cache regions by shed tool config~~

If we store the dogpile cache (a single ~ 130 MB file) on CVMFS I think we can really speedup first time bootstrapping of instances (I suppose CVMFS works better for large files).

xref https://github.com/galaxyproject/galaxy/issues/7752

### CODE REVIEW

**Comment by mvdbeek** [comment]
Fastest I've seen yet is a little less than 9 seconds (`galaxy.app INFO 2020-04-15 17:25:29,439 Galaxy app startup finished (8969.946 ms)`) with usegalaxy.org's  data tables and shed_tool_conf.xml from CVMFS. It's now more consistently around 9-10 seconds, which I attribute to creating hashes (where reading the files is the bottleneck) for tools and macros in the config watcher thread.

For web workers you'll have to add about 2 seconds for starting the API and web controllers.

This is the current profiling result:
```
*** PROFILER RESULTS ***
load_galaxy_app (./scripts/galaxy-main:82)
function called 1 times

         10165333 function calls (9740155 primitive calls) in 13.318 seconds

   Ordered by: cumulative time, internal time, call count
   List reduced from 13470 to 120 due to restriction <120>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000   13.340   13.340 galaxy-main:82(load_galaxy_app)
        1    0.000    0.000   10.854   10.854 app.py:63(__init__)
        1    0.000    0.000    7.678    7.678 __init__.py:990(_configure_toolbox)
        1    0.000    0.000    7.628    7.628 __init__.py:318(__init__)
        1    0.000    0.000    7.627    7.627 base.py:1201(__init__)
        1    0.000    0.000    7.615    7.615 base.py:78(__init__)
        1    0.000    0.000    7.261    7.261 base.py:134(_init_tools_from_configs)
        3    0.004    0.001    7.260    2.420 base.py:165(_init_tools_from_config)
4786/1426    0.013    0.000    7.169    0.005 base.py:248(load_item)
     1423    0.070    0.000    7.163    0.005 base.py:775(_load_section_tag_set)
     3321    0.061    0.000    6.878    0.002 base.py:641(_load_tool_tag_set)
     3446    0.014    0.000    6.457    0.002 base.py:853(load_tool)
     3446    0.009    0.000    6.191    0.002 __init__.py:371(create_tool)
     3446    0.005    0.000    4.084    0.001 <decorator-gen-3>:1(get_expanded_tool_source)
     3446    0.009    0.000    4.079    0.001 region.py:1351(get_or_create_for_user_func)
     3446    0.016    0.000    4.055    0.001 region.py:808(get_or_create)
     3446    0.002    0.000    3.972    0.001 lock.py:186(__enter__)
     3446    0.006    0.000    3.970    0.001 lock.py:83(_enter)
     3466    0.007    0.000    3.941    0.001 region.py:901(get_value)
     3466    0.037    0.000    3.920    0.001 __init__.py:133(get)
     3426    1.965    0.001    2.991    0.001 __init__.py:147(value_decode)
  2853/81    0.011    0.000    2.567    0.032 <frozen importlib._bootstrap>:966(_find_and_load)
  2846/74    0.008    0.000    2.566    0.035 <frozen importlib._bootstrap>:936(_find_and_load_unlocked)
  2262/74    0.007    0.000    2.560    0.035 <frozen importlib._bootstrap>:651(_load_unlocked)
  2147/74    0.005    0.000    2.558    0.035 <frozen importlib._bootstrap_external>:672(exec_module)
  2953/74    0.001    0.000    2.547    0.034 <frozen importlib._bootstrap>:211(_call_with_frames_removed)
 2949/285    0.046    0.000    2.547    0.009 {built-in method builtins.exec}
 1722/823    0.003    0.000    2.459    0.003 {built-in method builtins.__import__}
       34    0.001    0.000    2.433    0.072 __init__.py:3(<module>)
        1    0.000    0.000    2.329    2.329 app.py:1(<module>)
12707/5407    0.010    0.000    2.148    0.000 <frozen importlib._bootstrap>:997(_handle_fromlist)
     3426    0.004    0.000    2.098    0.001 __init__.py:389(_create_tool_from_source)
     3426    0.010    0.000    2.094    0.001 __init__.py:295(create_tool_from_source)
     3426    0.034    0.000    2.052    0.001 __init__.py:527(__init__)
     3426    0.120    0.000    1.258    0.000 __init__.py:778(parse)
      141    0.002    0.000    1.118    0.008 __init__.py:1(<module>)
        1    0.000    0.000    1.110    1.110 __init__.py:1055(_configure_tool_data_tables)
        3    0.010    0.003    1.110    0.370 __init__.py:104(load_from_config_file)
     5773    0.011    0.000    0.935    0.000 decoder.py:334(decode)
     3426    0.005    0.000    0.927    0.000 __init__.py:302(loads)
     5773    0.908    0.000    0.908    0.000 decoder.py:345(raw_decode)
        1    0.003    0.003    0.891    0.891 mapping.py:4(<module>)
23206/9342    0.013    0.000    0.766    0.000 {built-in method builtins.next}
     6932    0.016    0.000    0.756    0.000 file.py:215(_dbm_file)
     7662    0.015    0.000    0.647    0.000 __init__.py:1286(debug)
     7717    0.013    0.000    0.632    0.000 __init__.py:1421(_log)
     1694    0.004    0.000    0.613    0.000 __init__.py:242(from_elem)
     1694    0.004    0.000    0.609    0.000 __init__.py:326(__init__)
     1694    0.017    0.000    0.587    0.000 __init__.py:332(configure_and_load)
     3426    0.010    0.000    0.582    0.000 jobs.py:81(__init__)
    308/2    0.001    0.000    0.531    0.266 mapper.py:3174(configure_mappers)
     3427    0.017    0.000    0.530    0.000 collections.py:40(__init__)
   360/24    0.000    0.000    0.512    0.021 base.py:419(_inspect_mapped_class)
   350/14    0.000    0.000    0.512    0.037 base.py:434(class_mapper)
        1    0.000    0.000    0.512    0.512 mapper.py:1333(_configure_all)
      153    0.003    0.000    0.508    0.003 mapper.py:1935(_post_configure_properties)
10435/3503    0.003    0.000    0.503    0.000 contextlib.py:79(__enter__)
    35256    0.479    0.000    0.479    0.000 {built-in method posix.stat}
     7717    0.007    0.000    0.442    0.000 __init__.py:1446(handle)
     1648    0.065    0.000    0.438    0.000 __init__.py:507(extend_data_with)
     7717    0.015    0.000    0.432    0.000 __init__.py:1500(callHandlers)
        1    0.000    0.000    0.421    0.421 config_watchers.py:1(<module>)
     7717    0.012    0.000    0.417    0.000 __init__.py:852(handle)
1528/1386    0.001    0.000    0.414    0.000 interfaces.py:189(init)
  489/347    0.003    0.000    0.398    0.001 relationships.py:1981(do_init)
        1    0.000    0.000    0.388    0.388 __init__.py:40(<module>)
     7717    0.013    0.000    0.384    0.000 __init__.py:982(emit)
2984/2241    0.000    0.000    0.365    0.000 loading.py:35(instances)
        2    0.001    0.001    0.363    0.181 query.py:3232(all)
     10/8    0.001    0.000    0.361    0.045 __init__.py:103(load)
     10/8    0.000    0.000    0.360    0.045 constructor.py:47(get_single_data)
        8    0.000    0.000    0.359    0.045 __init__.py:574(application_stack_class)
        8    0.000    0.000    0.359    0.045 inspect.py:1499(stack)
        8    0.000    0.000    0.359    0.045 inspect.py:1471(getouterframes)
      110    0.000    0.000    0.359    0.003 inspect.py:1432(getframeinfo)
      110    0.001    0.000    0.353    0.003 inspect.py:760(findsource)
        3    0.001    0.000    0.348    0.116 loading.py:81(<listcomp>)
7316/2792    0.027    0.000    0.347    0.000 loading.py:509(_instance)
        1    0.000    0.000    0.346    0.346 cache.py:161(__init__)
        1    0.002    0.002    0.346    0.346 cache.py:174(rebuild)
    13433    0.334    0.000    0.346    0.000 {built-in method io.open}
        1    0.000    0.000    0.346    0.346 __init__.py:1071(_configure_datatypes_registry)
      279    0.011    0.000    0.343    0.001 inspect.py:714(getmodule)
     3466    0.010    0.000    0.341    0.000 __init__.py:53(open)
        1    0.009    0.009    0.332    0.332 base.py:407(_load_tool_panel)
        6    0.000    0.000    0.304    0.051 __init__.py:6(<module>)
     2753    0.014    0.000    0.301    0.000 base.py:340(__add_tool_to_tool_panel)
    17219    0.013    0.000    0.301    0.000 version.py:48(parse)
     3428    0.012    0.000    0.295    0.000 histories.py:40(__init__)
2563/1532    0.003    0.000    0.291    0.000 deprecations.py:115(warned)
    17219    0.113    0.000    0.287    0.000 version.py:271(__init__)
     2147    0.011    0.000    0.279    0.000 <frozen importlib._bootstrap_external>:743(get_code)
      489    0.004    0.000    0.275    0.001 relationships.py:2047(_setup_join_conditions)
    13600    0.011    0.000    0.272    0.000 genericpath.py:16(exists)
      489    0.004    0.000    0.267    0.001 relationships.py:2373(__init__)
        1    0.011    0.011    0.266    0.266 registry.py:87(load_datatypes)
     2681    0.003    0.000    0.262    0.000 posixpath.py:391(realpath)
10435/3503    0.011    0.000    0.261    0.000 contextlib.py:85(__exit__)
       19    0.000    0.000    0.254    0.013 __init__.py:64(register_postfork_function)
     2796    0.017    0.000    0.253    0.000 <frozen importlib._bootstrap>:870(_find_spec)
     3426    0.003    0.000    0.247    0.000 xml.py:263(parse_requirements_and_containers)
     3426    0.046    0.000    0.244    0.000 requirements.py:206(parse_requirements_from_xml)
    13750    0.015    0.000    0.242    0.000 hdas.py:43(__init__)
     3466    0.047    0.000    0.241    0.000 __init__.py:97(whichdb)
2779/2681    0.028    0.000    0.238    0.000 posixpath.py:400(_joinrealpath)
     3466    0.229    0.000    0.229    0.000 {method 'close' of '_gdbm.gdbm' objects}
     2796    0.002    0.000    0.224    0.000 <frozen importlib._bootstrap_external>:1149(find_spec)
2800/2796    0.007    0.000    0.222    0.000 <frozen importlib._bootstrap_external>:1117(_get_spec)
    15435    0.222    0.000    0.222    0.000 {method 'write' of '_io.TextIOWrapper' objects}
        1    0.000    0.000    0.221    0.221 __init__.py:934(configure_logging)
        1    0.000    0.000    0.221    0.221 config.py:800(dictConfig)
        1    0.000    0.000    0.221    0.221 config.py:493(configure)
        2    0.000    0.000    0.221    0.110 config.py:467(configure_custom)
        1    0.000    0.000    0.220    0.220 config.py:651(configure_formatter)
        1    0.000    0.000    0.219    0.219 __init__.py:597(application_stack_log_formatter)
     1648    0.047    0.000    0.218    0.000 __init__.py:513(parse_file_fields)
       45    0.018    0.000    0.217    0.005 __init__.py:241(parse_xml)
    13750    0.015    0.000    0.215    0.000 datasets.py:242(__init__)
     1627    0.008    0.000    0.213    0.000 __init__.py:425(merge_tool_data_table)
    31653    0.017    0.000    0.210    0.000 re.py:286(_compile)
```

**Comment by mvdbeek** [comment]
One thing I didn't think of is that delaying the tool parameter parsing until after the `fork` means higher memory consumption in each `forked` process as all those parameters are kept in memory. For main's toolbox that amounts to about 260 MB of memory. I guess I'll make this configurable, on on-demand instances you probably rarely ever finalize all the tools (you can easily do that though, by callings /api/tools/tests_summary for instance), so that would actually save some memory.
If you never fork (for instance you run handlers with `./scripts/galaxy-main`) it would also not lead to increased memory usage.

**Comment by mvdbeek** [comment]
Alright, this is ready for review!

File: doc/source/admin/galaxy_options.rst [hunk 1]
```diff
@@ -1002,6 +1002,31 @@
 :Type: str
 
 
+~~~~~~~~~~~~~~~~~~~~~~~
+``tool_cache_data_dir``
```
Comment (mvdbeek) [inline doc/source/admin/galaxy_options.rst:1006]: need to make the cache read-only

**Comment by mvdbeek** [comment]
I think this is ready. I was briefly tempted by introducing a `read_only` flag for the cache, but I realised that the per-shed config is going to be sufficient -- when an update on cvmfs occurs, the cache will be updated as well, and you can't update the tools or the shed_tool_conf file anyway, so this should be unnecessary (however a read-only attribute would probably be useful on the tool_conf.xml leve, but that's a different topic).

**Comment by jmchilton** [comment]
If it were anyone else I'd be terrified. Impressive work!

**Pull request merged**